modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java

```diff
@@ -855,19 +860,22 @@ public CompletableFuture<Ignite> start(Path configPath) {
                                     clientHandlerModule,
                                     deploymentManager,
                                     sql
-                            );
+                            ));
 
                             // The system view manager comes last because other components
                             // must register system views before it starts.
-                            lifecycleManager.startComponent(systemViewManager);
+                            startFuturesAfterJoin.add(lifecycleManager.startComponent(systemViewManager));
                         } catch (NodeStoppingException e) {
                             throw new CompletionException(e);
                         }
                     }, startupExecutor)
                     .thenComposeAsync(v -> {
                         LOG.info("Components started, performing recovery");
 
-                        return recoverComponentsStateOnStart(startupExecutor);
+                        return recoverComponentsStateOnStart(
+                                startupExecutor,
+                                CompletableFuture.allOf(startFuturesAfterJoin.toArray(CompletableFuture[]::new))
```

Comment:
Why do we need to pass this future here?

Reply:
This was done to satisfy the third item from the ticket

> IgniteImpl startup procedure should be updated to collect the futures from all components and wait for their completion inside recoverComponentsStateOnStart, after metaStorageMgr.notifyRevisionUpdateListenerOnStart() has been called, but before Watches are deployed (metaStorageMgr.deployWatches())

Reply:
Why don't we account for `startFuturesBeforeJoin` here then?

Reply:
We do, but when we return this future from `org.apache.ignite.internal.app.IgniteImpl#start`, there is chaining 
> return CompletableFuture.allOf(startFuturesBeforeJoin.toArray(CompletableFuture[]::new)).thenCompose((ignored) -> startFuture);

Reply:
Why do we wait for `startFuturesAfterJoin` before waiting for `startFuturesBeforeJoin` then? I'm confused.

Reply:
Where do we do that? We have chaining `startFuturesBeforeJoin` -> `startFuturesAfterJoin`

Reply:
As you said above:

```
IgniteImpl startup procedure should be updated to collect the futures from all components and wait for their completion inside recoverComponentsStateOnStart, after metaStorageMgr.notifyRevisionUpdateListenerOnStart() has been called, but before Watches are deployed (metaStorageMgr.deployWatches())
```

You current logic is: wait for `startFuturesAfterJoin` and then deploy the watches. `startFuturesBeforeJoin` are not taken into account.

> Where do we do that? We have chaining startFuturesBeforeJoin -> startFuturesAfterJoin

Yes, you've added the chaining but you start all futures in parallel, so this chaining is useless in this regard.